Create pod eviction concept topic

shannonxtreme opened the pull request:
## Description

Work for #14507

Created a new draft concept topic for Pod Eviction. Merged content from out-of-resource.md and eviction-policy.md. 

## Questions/comments



### Closed issues
* ~~[Line 37] - the control plane, the scheduler, or "the control plane or kube-controller manager"?~~
* ~~[Line 88-91] - What is the best way to describe `nodefs` and `imagefs`? [Related comment thread](https://github.com/kubernetes/website/pull/27739#discussion_r621370901).~~
* ~~[Line 96-100] - Has the kubelet deprecated garbage collection for eviction in DiskPressure conditions?~~
* ~~[Line 245-251] - should we reword to make it clear that only the pods' cgroup is checked for memory eviction? [Related comment](https://github.com/kubernetes/website/pull/27739#discussion_r621604375)~~
* ~~Verify that the Known Issues list and any areas marked for "future deprecation" are still valid.~~
* ~~https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/ has many conceptual sections which I suggest we move to this topic. That task topic doesn't actually have any "tasks". We should consider adding instructions in that topic for actually configuring the parameters, and move the conceptual info to the `Eviction Parameters` section in this concept topic.~~
* ~~Where does this topic belong? **Scheduling and Eviction** or **Cluster Administration**? [Related discussion thread](https://github.com/kubernetes/website/pull/27739#discussion_r621516396)~~
* ~~[Line 199-216] - does "garbage collect" replace "delete" in all four points?~~

/sig docs
/language en
/do-not-merge work-in-progress
/assign @sftim

sftim commented:
I agree that https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/ doesn't really explain a task. We could actually redirect it to the new page if we don't replace it.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,50 @@
+weight: 50
+---
+
+*Eviction* is the process by which kubelet proactively fails pods to reclaim
+resources for nodes.
```
You could use a glossary definition shortcode here, if you make sure there is a suitable definition in the glossary.
https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#glossary

See https://github.com/kubernetes/website/blame/a22aa8fd4feec091e736bca12b2f8776ade0ac94/content/en/docs/concepts/overview/components.md#L17 for an example of where this is done elsewhere.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,50 @@
+
+## How eviction works
+
+`kubelet` monitors resources like CPU, memory, disk space, and disk inodes on
```
BTW I'd write:
```suggestion
The kubelet monitors resources like CPU, memory, disk space, and filesystem inodes on
```

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,50 @@
+
+## What's next
+
+* Learn how to [Configure Out of Resource Handling](docs/tasks/administer-cluster/out-of-resource/)
+* Learn about [PodDisruptionBudgets](docs/tasks/run-application/configure-pdb/)
+* Learn about [Quality of Service (QoS)](docs/tasks/configure-pod-container/quality-service-pod/)
```
```suggestion
* Learn how to [Configure Out of Resource Handling](/docs/tasks/administer-cluster/out-of-resource/)
* Learn about [PodDisruptionBudgets](/docs/tasks/run-application/configure-pdb/)
* Learn about [Quality of Service](/docs/tasks/configure-pod-container/quality-service-pod/) (QoS)
```

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,50 @@
+This gracefully terminates the pods. If the pods are managed by a Deployment, the
+Deployment creates new pods for Kubernetes to schedule. 
+
+When evicting pods, `kubelet` respects your configured [PodDisruptionBudget](docs/tasks/run-application/configure-pdb/).
+You can control which pods are evicted last by setting the [QoS level](docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed)
```
```suggestion
When evicting pods, `kubelet` respects your configured [PodDisruptionBudget](/docs/tasks/run-application/configure-pdb/).
You can control which pods are evicted last by setting the [QoS level](/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed)
```

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,50 @@
+on those pods to `Guaranteed`.
+
+You can control the various parameters `kubelet` uses to make eviction decisions.
+For instructions, see [Configure Out of Resource Handling](docs/tasks/administer-cluster/out-of-resource/).
```
```suggestion
For instructions, see [Configure Out of Resource Handling](/docs/tasks/administer-cluster/out-of-resource/).
```

sftim reviewed (commented):
Some early feedback that I hope is useful

shannonxtreme commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,397 @@
+calculates the percentage value relative to the total capacity associated with
+the signal. 
+
+The value for `memory.available` is derived from the cgroupfs instead of tools
```
I'm not 100% sure what this means or if it is accurate. (line 59-68)

shannonxtreme commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,397 @@
+filesystems. Kubelet does not support other configurations. For example, do not
+store volumes and logs in a dedicated filesystem.
+
+{{<note>}}
```
Is this still a "future release" feature or has garbage collection been deprecated?

shannonxtreme commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,397 @@
+or the default to only allow `DaemonSet` pods to run when there are enough 
+resources.
+
+## Deprecation of existing feature flags to reclaim disk 
```
Is this still needed here?

sftim replied:
I would check https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ and see if the “existing flag” names still exist.

To be honest, I think it's OK to omit this section even if the deprecated command line flags are still available. The page at https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ is the best place to find what command line flags are current.

shannonxtreme commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,397 @@
+| `--low-diskspace-threshold-mb`             | `--eviction-hard` or `eviction-soft`    |
+| `--outofdisk-transition-frequency`         | `--eviction-pressure-transition-period` |
+
+## Known issues
```
Have these issues been fixed since this page was last updated?

sftim replied:
I don't know who's best to confirm that. This advice does look stale. Try asking in `#sig-node` on the Kubernetes Slack.

shannonxtreme commented:
I ported over content from the Out of Resource Handling topic (and made lots of edits to it). We can remove that topic entirely. Same for "Eviction Policy" - that content is now in Eviction.md.

sftim commented on content/en/docs/reference/glossary/eviction.md:
```diff
@@ -0,0 +1,22 @@
+title: Eviction
+id: eviction
+date: 2019-02-01
+full_link: /docs/concepts/scheduling-eviction/pod-eviction
```
nit:
```suggestion
full_link: /docs/concepts/scheduling-eviction/pod-eviction/
```

sftim commented on content/en/docs/reference/glossary/eviction.md:
```diff
@@ -0,0 +1,22 @@
+---
+title: Eviction
+id: eviction
+date: 2019-02-01
```
nit, very minor thing:
```suggestion
date: 2021-04-27
```

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,397 @@
+1. Pod resource usage relative to defined requests
+1. [Pod priority](/docs/concepts/configuration/pod-priority-preemption/)
+1. Resource consumption relative to scheduling requests
+1. Pod [QoS](/docs/tasks/configure-pod-container/quality-service-pod/)
```
```suggestion
1. Pod [QoS class](/docs/tasks/configure-pod-container/quality-service-pod/)
```
?

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,397 @@
+When the kubelet evicts pods to reclaim disk resources, it sorts pods by QoS and
+terminates pods using the largest amount of disk. When the kubelet responds to
+`inode` starvation, it uses the QoS to make eviction decisions. The kubelet sorts
+pods differently based on whether the node has a dedicated `imagefs` filesystem.
```
```suggestion
pods differently based on whether the node has a dedicated `imagefs` filesystem:
```

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,397 @@
+
+The default `eviction-minimum-reclaim` is `0` for all resources.
+
+## Scheduler
+
+The scheduler respects reported node conditions when making scheduling decisions.
+
+For example, if `MemoryPressure` node condition is active on a node, the scheduler
+does not schedule new `BestEffort` pods on the node. If the `DiskPressure` node
+condition is active, the scheduler does not schedule any new pods on the node. 
```
This section is not really about eviction.

sftim replied:
Maybe move it to https://kubernetes.io/docs/concepts/architecture/nodes/ or check that the content there already covers this?

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,397 @@
+
+The kubelet supports the following filesystems:
+
+1. `nodefs`: Volumes, daemon logs, etc.
+1. `imagefs`: Container runtimes use `imagefs` to store container images and
```
What are nodefs and imagefs? (For example, would I find these names in `/proc/mounts`?)

(I think these are symbolic names used within the kubelet and its configuration, and not used elsewhere). We should clarify what these names mean.

shannonxtreme replied:
`nodefs` and `imagefs` do seem like the symbolic names for the node's root filesystem (nodefs) and the runtime filesystem (imagefs), which in Docker is apparently "thin pool". 

I reworded these points to: 

1. `nodefs`: The node's root filesystem, used for local disk volumes, emptyDir, daemon logs, and more.
1. `imagefs`: An optional filesystem that container runtimes use to store images and writable layers.

sftim replied:
Further nit: when Kubernetes people say “node's root filesystem” they mean “filesystem containing `/var/lib/kubelet`” which may not be the filesystem that contains `/`.
I'm not sure if there's a good name for it though. “Node's root filesystem” is nice and short, just sometimes wrong.

shannonxtreme replied:
@dashpole or @bobbypage can you confirm?

dashpole replied:
The above are all correct.

shannonxtreme replied:
changed to 

1. `nodefs`: The node's root filesystem, used for local disk volumes, emptyDir,
   log storage, and more. For example, `nodefs` can provide `/` or `/var/lib/kubelet`. 

Does that cover potential confusion?

sftim replied:
Strictly speaking _nodefs_ refers to the filesystem containing the kubelet's main directory (that is, `/var/lib/kubelet`). If we can avoid writing `/` or mentioning “root” that avoids a possible point of confusion.

shannonxtreme replied:
Updated to 1. `nodefs`: The node's main filesystem, used for local disk volumes, emptyDir,
   log storage, and more. For example, `nodefs` contains `/var/lib/kubelet/`.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,397 @@
+
+{{< glossary_definition term_id="eviction" length="all" >}}
+
+During an eviction, the kubelet sets the `PodPhase` for the selected pods to `Failed`.
+This gracefully terminates the pods. If the pods are managed by a Deployment, the
+Deployment creates new pods for Kubernetes to schedule. 
```
```suggestion
During an eviction, the kubelet sets the `PodPhase` for the selected pods to `Failed`.
This gracefully terminates the pods.

If the pods are managed by a workload resource (such as
{{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}}
or {{< glossary_tooltip text="Deployment" term_id="deployment" >}}) that
replaces failed pods, the control plane creates new pods in place of the pod(s)
that were evicted.
```
?

shannonxtreme replied:
Done

shannonxtreme commented:
@sftim I've reached out to `#sig-node` and I'll re-request a review from you once I have the info I need from them 😄

ehashman commented:
/cc @dashpole @bobbypage

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+{{< glossary_definition term_id="eviction" length="all" >}}
+
+During an eviction, the kubelet sets the `PodPhase` for the selected pods to `Failed`.
+This gracefully terminates the pods. 
```
Note that it does not respect the pod's terminationGracePeriodSeconds, and uses a 0s grace-period.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+replaces failed pods, the control plane creates new pods in place of the evicted pods.
+
+{{<note>}}
+The kubelet attempts to [reclaim node-level resources](#reclaim-node-resources)
```
might be helpful to give an example, like removing unused images when under disk pressure.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+---
+title: Pod Eviction
```
Is this for the eviction api (the one used when draining a node), or out-of-resource eviction?

dashpole replied:
This seems to mix details of both.

dashpole replied:
It is also somewhat strange to have kubelet resource management under a scheduling-related section.

shannonxtreme replied:
This is meant to include all situations where eviction happens (so from what I understand, kubelet, draining, the eviction API?). The section is Scheduling and Eviction so I put it there - what's your recommendation?

dashpole replied:
I'm not entirely sure what scheduling and eviction have in common... They seem like two entirely separate topics.

shannonxtreme replied:
Hmm. What about under Concepts > Cluster Administration instead?

dashpole replied:
That seems like a reasonable place

sftim replied:
> I'm not entirely sure what scheduling and eviction have in common... They seem like two entirely separate topics.

Eviction quite often triggers scheduling (of the replacement). I'm in favor of presenting the two concepts alongside each other; in fact, I'd like to include preemption in the mix as well. That way we can contrast preemption with eviction and explain how the two differ - easier if the concepts are presented as siblings.

Also, people who aren't cluster administrators are likely to want to learn about eviction, especially the kind where the kubelet triggers it without prompting.

bobbypage replied:
+1 i would move this out of scheduling section and make it clear this is specifically for local kubelet eviction / out-of-resource eviction

sftim replied:
> I'm not entirely sure what scheduling and eviction have in common... They seem like two entirely separate topics.

Eviction quite often triggers scheduling (of the replacement). I'm in favor of presenting the two concepts alongside each other; in fact, I'd like to include preemption in the mix as well. That way we can contrast preemption with eviction and explain how the two differ - easier if the concepts are presented as siblings.

Also, people who aren't cluster administrators are likely to want to learn about eviction, especially the kind where the kubelet triggers it without prompting.

shannonxtreme replied:
I'm still a little finicky with github so I'll finish up edits to content before I move it 😝

shannonxtreme replied:
@sftim makes a good point as well - it made more sense to me under the scheduling section. I will hold off on moving the topic until we decide where it needs a home

alculquicondor replied:
We had a long discussion in #19081

After seeing this PR, 1/3 types of eviction directly relate to scheduling (currently the PR is not talking about preemption). The others are more about node health, which is under the kubelet domain.

Maybe a better home would be somewhere in https://kubernetes.io/docs/concepts/architecture/

But I know @sftim feels very strongly about this belonging within scheduling. So maybe it's is fine, as long as we change the URL from `scheduling-eviction` to `scheduling-and-eviction`. I thought we had done this already.

sftim replied:
> change the URL from `scheduling-eviction` to `scheduling-and-eviction`

We usually omit short words like “and” from URLs - for example, “Services And Networking” is represented in the URL as `services-networking`.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+before it terminates end-user pods.
+{{</note>}}
+
+When evicting pods, the kubelet respects your configured [PodDisruptionBudget](/docs/tasks/run-application/configure-pdb/).
```
This is false.  It does not respect PDBs in out-of-resource scenarios.  When using the [eviction API](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api), PDB is respected.  But it is not respected during kubelet evictions.

shannonxtreme replied:
So there's two main ways in which eviction is triggered - out-of-resource scenarios and the eviction API/draining a node? Would those fall under automated and manual eviction respectively?

dashpole replied:
Yes, but i'm trying to stress that "eviction" isn't one thing in kubernetes.  Its two completely separate things that were unfortunately given the same name.  But they share no characteristics except the name, and the fact that pods stop running where they are currently running.

shannonxtreme replied:
Okay. This topic will be the perfect spot to explain the distinction between the two types then, and to stress that they shouldn't and can't be interchangeable. I'll add content that tries to do that and request a review when I do!

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+{{</note>}}
+
+When evicting pods, the kubelet respects your configured [PodDisruptionBudget](/docs/tasks/run-application/configure-pdb/).
+You can control which pods are evicted last by setting the [Quality of Service class](/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed)
```
This is incorrect.  QoS has no impact on eviction ordering.  See https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#evicting-end-user-pods for what is taken into account.

shannonxtreme replied:
Thanks for the comments! 

The [Priority and Preemption page](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#interactions-of-pod-priority-and-qos) says that the only time Priority and QoS are both considered is during out-of-resource eviction. 

Also, the out of resource handling section has `BestEffort` and `Burstable` going before `Guaranteed`. Isn't that considering the QoS?

dashpole replied:
Yeah, we have been discussing that recently, as you aren't the only one confused by it.  We don't actually take QoS into account, but since QoS is based on setting requests, and requests relative to usage, you can reason from QoS to rough ordering of evictions.  For example, a Guaranteed pod is still going to be evicted last simply because it is impossible for its usage to exceed requests, which is the first criterion we rank by.  Suggestions would be welcome, but we have been discussing removing all references to QoS here.

dashpole replied:
We should update the Priority and Preemption page.  I wasn't aware of that section.  Note that the text after mentioning QoS where it talks about the algorithm doesn't mention QoS at all...  It is just the headline that is wrong there.

sftim replied:
@dashpole please feel very welcome to file an issue about any incorrect content you see!

shannonxtreme replied:
I created #27753 and added a PR for it. Accidentally didn't make the PR a draft, sorry.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+
+You can configure pod eviction to happen automatically, triggered by a controller. 
+
+You can also request pod eviction manually using the `kubectl drain` command
```
This is different from a kubelet eviction.

shannonxtreme replied:
To clarify, this topic is for all types of eviction. The [Safely Drain a Node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#use-kubectl-drain-to-remove-a-node-from-service) page states that `kubectl drain` evicts pods in a node, so I added that here as one of the possible methods.

dashpole replied:
Nearly all of this page does not apply to draining a node.  The eviction API does graceful deletion (which eviction is not), and _does_ respect pod disruption budgets, for example.  Your pod also won't be set to phase == Failed during a drain.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+1. `imagefs`: An optional filesystem that container runtimes use to store container
+   images and container writable layers.
+
+Kubelet auto-discovers these filesystems using cAdvisor and ignores other
```
cAdvisor is an implementation detail that only applies to linux.  I'd suggest not mentioning it here.

shannonxtreme replied:
Done.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+1. Pod resource usage relative to defined requests
+1. [Pod priority](/docs/concepts/configuration/pod-priority-preemption/)
+1. Resource consumption relative to scheduling requests
+1. Pod [QoS class](/docs/tasks/configure-pod-container/quality-service-pod/)
```
This isn't correct.  It is just the first 3 items in this list.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+
+The kubelet uses the following parameters to determine pod eviction order:
+
+1. Pod resource usage relative to defined requests
```
This reads as being the same as number 3.  Suggestion: "If pod resource usage exceeds defined requests or not"

shannonxtreme replied:
Out of curiosity, what's the difference between 1 and 3?

dashpole replied:
1. Requests < Usage ?  True or false
2. ...
3. Usage - Requests

shannonxtreme replied:
OH. Okay, point 3 should be clearer - by how much does usage exceed requests?

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+1. Resource consumption relative to scheduling requests
+1. Pod [QoS class](/docs/tasks/configure-pod-container/quality-service-pod/)
+
+As a result, kubelet ranks and evicts pods in the following order:
```
This only applies to CPU and memory, as QoS doesn't take ephemeral storage requests into account.

shannonxtreme replied:
So something like "Note: the QoS class ranking does not apply to ephemeral storage requests like `inodes` and `PIDs`."?

dashpole replied:
I'd suggest framing this section differently, if we keep it at all.  The QoS descriptions are meant to help users reason about behavior, rather than completely describe the algorithm.
```
For example, if a node is under MemoryPressure, it would evict pods in the following order:

1. BestEffort or Burstable pods are the only pods where memory usage is allowed to exceed requests.  Among pods where usage is greater than requests, those with the lowest Priority, and then those using the most in excess of their requests are evicted first.
2. `Guaranteed` pods may not exceed their requests, since their requests and limits are equal.  Burstable pods may also not be exceeding their requests when MemoryPressure is triggered.  These pods are the last to be evicted since they are operating within their requested resources.

Note: QoS is not actually used during ranking, but can help illustrate which pods are most likely to be evicted.  QoS does not take EphemeralStorage requests into account, so the above scenario won't apply if the node is under DiskPressure, for example.
```

shannonxtreme replied:
I like that suggestion a lot, especially based on your other comment about potentially removing QoS at some point. I'll rework.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+and to limit the impact of resource starvation on other Pods. In this case, it
+will choose to evict pods of lowest Priority first.
+
+When the kubelet evicts pods to reclaim disk resources, it sorts pods by QoS and
```
This isn't correct.  It is by the same algorithm as CPU and memory.  Note that there are no requests for inodes or PIDs, so priority ends up being the first criterion.

shannonxtreme replied:
So for disk resources, it uses the same logic as for CPU and memory with the exception of using the QoS class? And then for `inodes` and `PIDs`, only the Priority is used to rank the pods?

dashpole replied:
For disk, inodes, and PIDs, it uses the same logic as for memory.  (There isn't CPU eviction).  It does _not_ ever take QoS into account.  There are no requests for inodes or PIDs, so the algorithm:

1. Usage > request ?
2. Priority
3. Usage - request

Becomes 

1. Usage > 0 ? (always true)
2. Priority
3. Usage - 0 = Usage

For inodes and PIDs.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+
+The `kubelet` currently polls `cAdvisor` to collect memory usage stats at a regular interval. If memory usage
+increases within that window rapidly, the `kubelet` may not observe `MemoryPressure` fast enough, and the `OOMKiller`
+will still be invoked. We intend to integrate with the `memcg` notification API in a future release to reduce this
```
memcg integration is done, but is not enabled by default, IIRC.  --kernel-memcg-notification is the flag to enable it on the kubelet.

shannonxtreme replied:
Is the workaround on line 357 still applicable?

dashpole replied:
Using Node allocatable is probably a better solution. (by setting kube-reserved and system-reserved)

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+
+### kubelet may evict more Pods than needed
+
+The Pod eviction may evict more Pods than needed due to stats collection timing gap. This can be mitigated by adding
```
This is done

shannonxtreme commented on content/en/docs/reference/glossary/eviction.md:
```diff
@@ -0,0 +1,22 @@
+- Operation
+- Fundamental
+---
+Eviction is the process by which the kubelet proactively fails pods to reclaim
```
@dashpole this definition isn't correct, is it? From what you've said in your reviews, it isn't *just* the kubelet that evicts pods, the API does too? 

So would the definition be more correct as: "Eviction is the process by which pods are terminated proactively to reclaim resources for nodes", and then in the <!-- more --> section mention that eviction can happen automatically through kubelet or manually through the Eviction API?

dashpole replied:
I would say "Out-of-resource Eviction is the process by which pods are terminated proactively to reclaim resources for nodes"

You can also say: "If you want to delete a pod, but respect policy controls during deletion, you can create an `Eviction` object, which will gracefully terminate your pod".

If you only want to talk about the former (out-of-resource evictions), then your definition is correct.  If you want to include the latter, you would need to make the language more general to the point where it wouldn't mean much to users.

shannonxtreme replied:
It might be a good idea to separate the two in the glossary - "Out-of-resource eviction" and "Eviction object". That way we can include both glossary terms in our concept topic and make a clearer distinction between the two.

sftim replied:
> eviction object

I'd call that an eviction resource. The difference is subtle but I'd be confident that creating an eviction means creating a resource, whereas different folks might or might not agree that the thing you create is an object.

sftim replied:
(the downside though is that “resource” is also the thing you can be short of when the kubelet evicts your Pod. That's, uh, a ~bug~ feature of English, though, and not in our power to adjust).

sftim replied:
:+1: for two glossary definitions. Have a think about what the short names for these should be.

sftim replied:
“graceful Pod deletion” could be the term for the thing you do when draining a node / creating a Pod eviction using the API

shannonxtreme reviewed (commented):
Implemented some of @dashpole's edits and added followups to others.

bobbypage commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+If the node only has a `nodefs` filesystem that meets eviction thresholds,
+the kubelet frees up disk space in the following order:
+
+1. Delete dead pods and containers
```
"Garbage collect" might be more appropriate here

shannonxtreme replied:
Just for 1. or for both?

shannonxtreme replied:
ping @bobbypage

bobbypage commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+`system-reserved` or `kube-reserved` allocations, and the node only has
+`Guaranteed` or `Burstable` pods using less resources than requests left on it,
+then the kubelet must choose to evict one of these pods to preserve node stability
+and to limit the impact of resource starvation on other Pods. In this case, it
```
I'm not sure regarding the `system daemon` section. 

@dashpole to keep me honest, but as far as I'm aware kubelet eviction for memory is only looking at the kubepods cgroup `statsapi.SystemContainerPods`: https://github.com/kubernetes/kubernetes/blob/fbc93bd34c51ba81aebf789da67bfa497fa9225a/pkg/kubelet/eviction/helpers.go#L672 

So I don't think memory usage outside the pod cgroup (i.e. those by system daemons) would influence kubelet memory eviction.

shannonxtreme replied:
Thanks! I'll wait for confirmation  and then add it.

dashpole replied:
There are two memory eviction thresholds.

1. Sum(Pods) < Allocatable
2. All usage (including pods, system daemons) < Capacity - eviction threshold

System daemon usage isn't limited by kube-reserved in most setups, and comes into play for the second threshold.

shannonxtreme replied:
Leaving as-is for now.

bobbypage replied:
ah got it, thanks @dashpole for clarification. For #2 that checks is from here looks like: https://github.com/kubernetes/kubernetes/blob/fbc93bd34c51ba81aebf789da67bfa497fa9225a/pkg/kubelet/eviction/helpers.go#L666-L670 (i.e. `summary.Node.Memory`)

bobbypage commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
````diff
@@ -0,0 +1,375 @@
+For example, the following configuration sets minimum reclaim amounts: 
+
+```
+--eviction-hard=memory.available<500Mi,nodefs.available<1Gi,imagefs.available<100Gi
````
we likely want to point to the kubelet config variant of this config rather than the flags.

shannonxtreme replied:
Do you mean the YAML config? Where would I get my hands on that? ✋🏽

bobbypage replied:
yes, as defined here https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/

for example:

```
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
evictionHard:
    memory.available:  "200Mi"
```

full list of options are defined here: https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/

bobbypage commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+
+## Node out of memory behavior
+
+If the node experiences a system out of memory (OOM) event prior to the `kubelet`
```
I would remove "system" and just say out of memory (OOM).

There can be either cgroup or system OOMs. Most likely case is a cgroup OOM, e.g. a burstable pod which exceeds memory limit of node allocatable (top level pod cgroup) which then triggers a cgroup OOM.

sftim commented on content/en/docs/reference/glossary/eviction.md:
```diff
@@ -0,0 +1,22 @@
+  resources for nodes.
+aka:
+tags:
+- Operation
+- Fundamental
```
```suggestion
- operation
- fundamental
```

shannonxtreme commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,375 @@
+
+If the pods are managed by a workload resource (such as {{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}}
+or {{< glossary_tooltip text="Deployment" term_id="deployment" >}}) that
+replaces failed pods, the control plane creates new pods in place of the evicted pods.
```
What about "the scheduler creates..."?

shannonxtreme replied:
Replaced with "the control plane or `kube-controller-manager`"

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,402 @@
+with manual eviction. Do not use the two eviction types interchangeably. 
+{{</caution>}}
+
+## Manual eviction
```
I'd suggest "Graceful Eviction", as that is the primary distinction.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,402 @@
+  * Eviction signals
+  * Eviction thresholds
+  * Monitoring intervals
+  * Node conditions
```
nit: node conditions are an _output_ of eviction, rather than an input

dashpole commented:
Contents look correct to me.

shannonxtreme commented:
Hi @sftim @bobbypage @dashpole. I've compiled the currently outstanding questions/comments in the PR description (and below).

### Questions/comments

* Where does this topic belong? **Scheduling and Eviction** or **Cluster Administration**? [Related discussion thread](https://github.com/kubernetes/website/pull/27739#discussion_r621516396)
* [Line 37] - the control plane, the scheduler, or "the control plane or kube-controller manager"?
* [Line 88-91] - What is the best way to describe `nodefs` and `imagefs`? [Related comment thread](https://github.com/kubernetes/website/pull/27739#discussion_r621370901). 
* [Line 96-100] - Has the kubelet deprecated garbage collection for eviction in DiskPressure conditions? 
* [Line 199-216] - does "garbage collect" replace "delete" in all four points?
* [Line 245-251] - should we reword to make it clear that only the pods' cgroup is checked for memory eviction? [Related comment](https://github.com/kubernetes/website/pull/27739#discussion_r621604375)

After those are resolved, I'll go through the topic to clean up and add quality of life improvements and request a final review from Tim. Thanks 😄

dashpole commented:
Garbage collection is deprecated, but I think most providers still enable it.  I don't think its going away anytime soon.

shannonxtreme commented:
We should talk about potentially deleting out-of-resource-handling and eviction-policy since the content is here now. Not sure what the process is - is a redirect entry needed?

shannonxtreme commented:
My 2 cents: we add the topic location discussion to a separate meeting (maybe the next sig-docs meeting? Not sure how it works) and move it if needed based on that, since it may also bring up moving preemption.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,414 @@
+
+{{<glossary_definition term_id="graceful-eviction" length="all">}}
+
+Graceful evictions respect your configured [`PodDisruptionBudgets`](/docs/tasks/run-application/configure-pdb/). 
```
and also pod termination grace periods.

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,414 @@
+
+#### kubelet may not observe memory pressure right away
+
+The `kubelet` currently polls `cAdvisor` to collect memory usage stats at a regular interval. If memory usage
```
We now have memcg notifications implemented, so change this to:

"By default, the `kubelet` polls ..."

dashpole reviewed (approved):
Looks good!

bobbypage commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+being able to reclaim memory, the node depends on the [oom_killer](https://lwn.net/Articles/391222/)
+to respond.
+
+The kubelet sets an `oom_score_adj` value for each container based on the QoS for the pod.
```
Might be worth to note somewhere that in addition to logic here if the pod has priority `system-node-critical` there is special case of it getting -997 score: 

https://github.com/kubernetes/kubernetes/blob/b58a7e2/pkg/kubelet/qos/policy.go#L41-L44

sftim replied:
(that could come in a follow-up PR, BTW)

shannonxtreme replied:
Lol, yes, I am going to make new suggestions in new PRs (but I implemented this one already).

bobbypage commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+
+| Quality of Service | oom_score_adj                                                                     |
+|--------------------|-----------------------------------------------------------------------------------|
+| `Guaranteed`       | -998                                                                              |
```
nit, but the code seems to set this to -997, not -998: https://github.com/kubernetes/kubernetes/blob/b58a7e2/pkg/kubelet/qos/policy.go#L30

it was changed in https://github.com/kubernetes/kubernetes/pull/71269

shannonxtreme replied:
So it's the same score as for system-node-critical priority pods?

bobbypage replied:
yes, exactly: https://github.com/kubernetes/kubernetes/blob/b58a7e233e06f5e4b58a637af8e36cb4f59c001b/pkg/kubelet/qos/policy.go#L41-L44

dashpole commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+in a {{<glossary_tooltip text="node" term_id="node" length="short">}} are
+terminated. 
+
+In Kubernetes, there are two types of pod eviction:
```
nit: I don't think its the apiserver that controls graceful eviction.  Its probably some controller in the kube-controller-manager, but i'd double-check.

sftim replied:
> Graceful eviction is initiated by api-server

It's some _client_ of the API server (eg `kubectl drain`) that creates these Eviction subresources of Pod.

sftim replied:
The API client could also be a controller, whether that happens within the cluster's control plane or outside it.

shannonxtreme replied:
so the Eviction object is created by interacting with the Eviction API through a client of apiserver like kubectl drain or a controller. When created, the apiserver (?) triggers graceful termination?

bobbypage replied:
ack, thanks for clarifying @dashpole and @stfim ; it's what I was trying to across - a client explicitly calls API server to create eviction object as opposed to kubelet out-of-resource eviction (which does not involve any external client creating eviction via API server).

Regarding what actually deletes pod during an eviction, I took a look and it appears it actually is api-server ([here](https://github.com/kubernetes/kubernetes/blob/2bda8ac3ec163ef158ca45d4f49de857c5107070/pkg/registry/core/pod/storage/eviction.go#L175-L180)). If the eviction cannot be performed due to outstading PDB it appears API-server will deny creation of the eviction object (and return `429` HTTP status) ([xref](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api)).

alculquicondor commented on content/en/docs/concepts/scheduling-eviction/kube-scheduler.md:
```diff
@@ -66,6 +66,11 @@ resource requests. After this step, the node list contains any suitable
+The scheduler also respects [node conditions](/docs/concepts/scheduling-eviction/pod-eviction#node-conditions)
+when making scheduling decisions. For example, if the `MemoryPressure` node condition is active on a node, the scheduler
+does not schedule new `BestEffort` pods on the node. If the `DiskPressure` node
+condition is active, the scheduler does not schedule any new pods on the node. 
```
Well, the fact that the admission plugin doesn't add a toleration means that nodes that have the taint would get no pods scheduled. The special case is rather `MemoryPressure`, which has an explicit toleration for non `BestEffort` pods.

Now.... it's not really the scheduler which implements this. It's the apiserver. We can add this detail in https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-nodes-by-condition, which already explains how taints and node conditions are related.

shannonxtreme replied:
So in this case, remove the `DiskPressure` part?

We should create a new issue to update Taints and Tolerations as well.

alculquicondor replied:
The statement is correct, but I would change the order.

> For example, If the `DiskPressure` node condition is active, the system doesn't allow any new Pods to schedule on the node. If the `MemoryPressure` node condition is active on a node, the system doesn't allow new `BestEffort` pods on the node.

BUT, I wouldn't have this in under the kube-scheduler page (because it's not something kube-scheduler does). It should be in the taints and tolerations doc.

shannonxtreme replied:
Opened #27878 for this.

alculquicondor replied:
Can you remove this paragraph from here then?

shannonxtreme replied:
Yup already removed. I'm holding off from pushing until other comments are resolved. Anything I resolve here is fixed in the working copy :)

alculquicondor commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+
+{{<caution>}}
+Out-of-resource eviction does not share any characteristics
+with graceful eviction. Do not use the two eviction types interchangeably. 
```
You don't "use" out-of-resource eviction. The system applies it.

shannonxtreme replied:
I'll remove that sentence. Maybe "Out-of-resource eviction and graceful eviction are not interchangeable"?

alculquicondor replied:
I think the caution is unnecessary.

alculquicondor commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+In Kubernetes, there are two types of pod eviction:
+
+  1. Graceful eviction
+  2. Out-of-resource eviction ({{<glossary_tooltip text="kubelet" term_id="kubelet" length="short">}}
```
There are 2 types of out-of-resource eviction:

- The first is more like node resource-pressure eviction, acted on by the kubelet.
- The second is cluster level out-of-resource preemption, which clears space from any node to be able to schedule higher priority pods. This is acted on by scheduler.

shannonxtreme replied:
What's the difference between eviction and preemption? We have a separate topic talking about preemption. 

For this I can add: 
There are two types of out-of-resource eviction: 
* Node-level eviction: managed by the kubelet and triggered by resource pressure in nodes
* Cluster-level preemption: managed by the scheduler and triggered to clear space on nodes to schedule higher priority pods. For more information, see [Pod Priority and Preemption](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#preemption).

alculquicondor replied:
lgtm, but maybe @sftim has a better phrasing.

shannonxtreme replied:
@sftim - I updated the glossary entry for out-of-resource eviction. I'll work it into the doc in a way that makes sense and get you to look at it on my next push.

shannonxtreme commented:
@alculquicondor @bobbypage I added followups to your review comments. Take a look when you can and let me know what you think. I'll hold off on pushing till we're resolved :)

sftim commented on content/en/docs/reference/glossary/kubelet-eviction.md:
```diff
@@ -0,0 +1,22 @@
+date: 2021-04-27
+full_link: /docs/concepts/scheduling-eviction/pod-eviction/
+short_description: >
+  Out-of-resource eviction is the process by which the kubelet proactively fails
+  pods to reclaim resources for nodes.
```
```suggestion
  The process by which the kubelet proactively fails pods in order to reclaim resources.
```

sftim commented on content/en/docs/reference/glossary/graceful-eviction.md:
```diff
@@ -0,0 +1,20 @@
+- operation
+- fundamental
+---
+Graceful eviction is the process by which you use the [Eviction API](/docs/reference/generated/kubernetes-api/v1.21/#create-eviction-pod-v1-core)
```
```suggestion
Graceful eviction is the process by which you use the [Eviction API](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#create-eviction-pod-v1-core)
```
this then tracks the version we're documenting.

sftim commented on content/en/docs/reference/glossary/graceful-eviction.md:
```diff
@@ -0,0 +1,20 @@
+aka:
+tags:
+- operation
+- fundamental
```
I'd omit this one from `fundamental`.

sftim commented on content/en/docs/reference/glossary/graceful-eviction.md:
```diff
@@ -0,0 +1,20 @@
+date: 2021-04-27
+full_link: /docs/concepts/scheduling-eviction/pod-eviction/
+short_description: >
+  Graceful eviction is the process by which you create an `Eviction` object using
```
I'm not sure that backticks (and other Markdown formatting) in a short description will work how you might expect. I'd remove them unless we're confident they always render OK.

sftim commented on content/en/docs/reference/glossary/graceful-eviction.md:
```diff
@@ -0,0 +1,20 @@
+title: Graceful eviction
+id: graceful-eviction
+date: 2021-04-27
+full_link: /docs/concepts/scheduling-eviction/pod-eviction/
```
Maybe add a fragment identifier to the end of this URL?a

sftim commented on content/en/docs/reference/glossary/kubelet-eviction.md:
```diff
@@ -0,0 +1,22 @@
+title: Out-of-resource eviction
+id: kubelet-eviction
+date: 2021-04-27
+full_link: /docs/concepts/scheduling-eviction/pod-eviction/
```
Maybe add a fragment identifier to the end of this URL?

sftim commented on static/_redirects:
```diff
@@ -91,7 +91,8 @@
 /docs/concepts/cluster-administration/network-plugins/     /docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/ 301
-/docs/concepts/cluster-administration/out-of-resource/     /docs/tasks/administer-cluster/out-of-resource/ 301
+/docs/concepts/cluster-administration/out-of-resource/     /docs/concepts/scheduling-eviction/pod-eviction/ 301   
+/docs/tasks/administer-cluster/out-of-resource/     /docs/concepts/scheduling-eviction/pod-eviction/ 301
```
This redirect should ideally be grouped with other tasks-section redirects.

shannonxtreme replied:
Ah, sorry, I was grouping the actual topics together. I'll move it.

sftim reviewed (commented):
More feedback, nothing really binding though.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+  2. Out-of-resource eviction ({{<glossary_tooltip text="kubelet" term_id="kubelet" length="short">}}
+     eviction)
+
+{{<caution>}}
```
This should be a `{{< note >}}` callout rather than a `{{< caution >}}`; we reserve `{{< caution >}}` for things that are a bit more serious (eg: potential for security problems, but only if you take specific steps). 

There is also `{{< warning >}}` for things you really ought to know (eg: don't use an allow-all authorization configuration except for local testing)

shannonxtreme replied:
Removed the note entirely.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+The kubelet does not respect your configured `PodDisruptionBudget` or the pod's
+`terminationGracePeriodSeconds`, and uses a `0s` grace period for termination.
+
+If the pods are managed by a {{<glossary_tooltip text="workload" term_id="workload" length="short">}}
```
```suggestion
If the pods are managed by a {{<glossary_tooltip text="workload" term_id="workload">}}
```
(`length` isn't relevant to tooltips; these either use the default text or the text you explicitly use as an override)

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+In Kubernetes, there are two types of pod eviction:
+
+  1. Graceful eviction
+  2. Out-of-resource eviction ({{<glossary_tooltip text="kubelet" term_id="kubelet" length="short">}}
```
```suggestion
  2. Out-of-resource eviction ({{<glossary_tooltip text="kubelet" term_id="kubelet" >}}
```

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+
+{{<note>}}
+Some kubelet garbage collection features are deprecated in favor of eviction.
+For a list of the deprecated features, see [kubelet garbage collection deprecation](http://localhost:1313/docs/concepts/cluster-administration/kubelet-garbage-collection/#deprecation)
```
This link is incorrect - make it a site-relative URL.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,417 @@
+1. Garbage collect dead pods and containers
+1. Delete unused images
+
+### Evicting end-user pods
```
```suggestion
### Pod selection for kubelet eviction
```
?

shannonxtreme commented:
@sftim - major items for review include the glossary entries and the **Out-of-resource eviction** overview section. 
@bobbypage - Could you please take another look at the kubelet-eviction glossary entry? In addition to an overall check.

bobbypage commented:
thanks @shannonxtreme, one small comment about glossary

alculquicondor commented on content/en/docs/reference/glossary/kubelet-eviction.md:
```diff
@@ -0,0 +1,26 @@
+
+There are two types of out-of-resource eviction: 
+
+1. Cluster-level preemption: the scheduler evicts pods from nodes in a cluster
```
Yes, have a separate entry for preemption.

shannonxtreme replied:
We already have a separate entry for preemption. I wonder if we should have link to that entry from the kubelet eviction entry, something like "To learn about cluster-level eviction, see preemption." Thoughts?

alculquicondor replied:
Let's figure something out first:
Do we want to consider kubelet eviction and preemption both as types of "out-of-resource eviction"? If so, the title of this should have 3 entries: out-of-resource eviction, node eviction (or kubelet eviction) and cluster eviction (or preemption). The first references the other 2.

If not, we just have 2 entries: node eviction and cluster eviction.

@sftim what is your take?

shannonxtreme replied:
Are we married to calling it "out-of-resource eviction"? If this was called kubelet eviction, cluster-level called preemption, and the Eviction object one called graceful eviction, would that make it cleaner?

shannonxtreme replied:
> Let's figure something out first:
> Do we want to consider kubelet eviction and preemption both as types of "out-of-resource eviction"? If so, the glossary should have 3 entries: out-of-resource eviction, node eviction (or kubelet eviction) and cluster eviction (or preemption). The first references the other 2.
> 
> If not, we just have 2 entries: node eviction and cluster eviction.

I guess since preemption isn't really "out-of-resource" in the same way that node eviction is (at least the way I, a relative newbie, understand it), I'm in favour of just having them as node eviction and cluster eviction. I can do that in this PR and then if we decide we do want them grouped, we can just add that in a future PR.

shannonxtreme replied:
Also - node-level eviction (so that there's nothing to suggest that entire nodes are getting booped)?

shannonxtreme commented on content/en/docs/reference/glossary/graceful-eviction.md:
```diff
@@ -0,0 +1,20 @@
+date: 2021-04-27
+full_link: /docs/concepts/scheduling-eviction/pod-eviction/#graceful-eviction
+short_description: >
+  Graceful eviction is the process by which use the Eviction API to create an
```
:( I blame Friday. Thanks for the catch!

sftim commented on content/en/docs/reference/glossary/preemption.md:
```diff
@@ -6,7 +6,7 @@ full_link: /docs/concepts/configuration/pod-priority-preemption/#preemption
   Preemption logic in Kubernetes helps a pending Pod to find a suitable Node by evicting low priority Pods existing on that Node.
 
-aka:
+aka: cluster-level eviction
```
I wouldn't characterize “preemption” as “cluster-level eviction” - I don't think many folks familiar with Kubernetes would see those two things as equivalent.

sftim commented on content/en/docs/reference/glossary/kubelet-eviction.md:
```diff
@@ -0,0 +1,22 @@
+---
+title: Node-level eviction
```
```suggestion
title: Node-initiated eviction
```
?

alculquicondor replied:
I think I prefer to highlight the reason, rather than what executes it. So my recommendation is Node-pressure eviction. Although I prefer Node-initiated than Node-level.

sftim commented on content/en/docs/reference/glossary/graceful-eviction.md:
```diff
@@ -0,0 +1,20 @@
+---
+title: Graceful eviction
+id: graceful-eviction
```
I want to use the word “exogenous”, but it's too complicated.
```suggestion
title: API-initiated eviction
```
How about this?

sftim commented on content/en/docs/reference/glossary/graceful-eviction.md:
```diff
@@ -0,0 +1,20 @@
+
+<!--more-->
+
+You can request graceful eviction either by directly calling the Eviction API 
+using a client of the kube-apiserver, like the `kubectl drain` command. 
+When an `Eviction` object is created, the API server terminates the Pod. 
```
```suggestion
You can request graceful eviction either by directly calling the Eviction API 
using a client of the kube-apiserver, like the `kubectl drain` command. 
When an `Eviction` object is created, the API server terminates the Pod.

API-initiated eviction is not the same as
[node-initiated eviction](/docs/concepts/scheduling-eviction/pod-eviction/#kubelet-eviction).
```

sftim commented on content/en/docs/reference/glossary/kubelet-eviction.md:
```diff
@@ -0,0 +1,22 @@
+
+<!--more-->
+
+The kubelet monitors resources like CPU, memory, disk space, and filesystem 
+inodes on your cluster's nodes. When one or more of these resources reach
+specific consumption levels, the kubelet can proactively fail one or more pods
+on the node to reclaim resources and prevent starvation. 
```
```suggestion
The kubelet monitors resources like CPU, memory, disk space, and filesystem 
inodes on your cluster's nodes. When one or more of these resources reach
specific consumption levels, the kubelet can proactively fail one or more pods
on the node to reclaim resources and prevent starvation. 

API-initiated eviction is not the same as
[node-initiated eviction](/docs/concepts/scheduling-eviction/pod-eviction/#kubelet-eviction).
```

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,421 @@
+---
+title: Pod Eviction
```
```suggestion
title: Eviction
```
You can't evict (or schedule, or preempt, or deschedule) anything other than a pod.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,421 @@
+
+In Kubernetes, there are two types of pod eviction:
+
+  1. Graceful eviction
```
```suggestion
  1. [Graceful eviction](#graceful-eviction)
  2. [Node-level eviction](#kubelet-eviction) ({{< glossary_tooltip text="kubelet" term_id="kubelet" >}}
     eviction)
```
_except_ that I'd also rename both terms; see feedback elsewhere.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,421 @@
+
+## Graceful eviction {#graceful-eviction}
+
+{{< glossary_definition term_id="graceful-eviction" length="all" >}}
```
Maybe just use the short form, and then add extra text that's similar to the rest of the glossary definition.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,421 @@
+
+## Node-level eviction {#kubelet-eviction}
+
+{{< glossary_definition term_id="kubelet-eviction" length="all" >}}
```
Maybe just use the short form, and then add extra text that's similar to the rest of the glossary definition.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,421 @@
+
+{{< glossary_definition term_id="kubelet-eviction" length="all" >}}
+
+Pod eviction can also happen at the cluster level during [pod preemption](/docs/concepts/configuration/pod-priority-preemption/#preemption).
```
I disagree. Preemption is not eviction. We should instead clarify that preemption is a different mechanism by which the control plane stops a potentially healthy Pod.

alculquicondor replied:
+1, it should have it's own subtitle.

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,421 @@
+
+{{<note>}}
+Some kubelet garbage collection features are deprecated in favor of eviction.
+For a list of the deprecated features, see [kubelet garbage collection deprecation](/docs/concepts/cluster-administration/kubelet-garbage-collection/#deprecation)
```
nit:
```suggestion
For a list of the deprecated features, see [kubelet garbage collection deprecation](/docs/concepts/cluster-administration/kubelet-garbage-collection/#deprecation).
```

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,421 @@
+Unlike pod eviction, if a container is OOM killed, the `kubelet` can restart it 
+based on its `RestartPolicy`.
+
+### Best practices
```
I'd be tempted to:
```suggestion
### Best practices {#kubelet-eviction-good-practice}
```

sftim commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,421 @@
+
+* Learn about [PodDisruptionBudgets](/docs/tasks/run-application/configure-pdb/)
+* Learn about [Quality of Service](/docs/tasks/configure-pod-container/quality-service-pod/) (QoS)
+* Check out the [Eviction API](/docs/reference/generated/kubernetes-api/v1.21/#create-eviction-pod-v1-core)
```
We should set the version from code:
```suggestion
* Check out the [Eviction API](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#create-eviction-pod-v1-core)
```

shannonxtreme commented:
@sftim @alculquicondor I implemented many of your comments. 

* Renamed node-level eviction to node pressure eviction and graceful eviction to API-initiated. 
* Added params to API links.
* Added preemption line in eviction overview (take a look - I don't think it needs its own subtitle).
* Changed the concept title to "eviction" from "pod eviction"

shannonxtreme commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,432 @@
+---
+title: Eviction
```
It's supposed to - it used to be Pod Eviction but as Tim said, only Pods can be evicted so I changed it to just Eviction :P

alculquicondor commented:
From the SIG Docs meeting, there are basically 2 proposals of concepts hierarchy. I'm not necessarily discussing the exact titles or how they are laid. It's more of a question of how the concepts are related to each other.

Option 1 (somewhat the current state of this PR):
- Scheduling, Preemption and Eviction:
  - Other scheduling topics...
  - Preemption
  - Eviction
    - User initiated Eviction
    - Kubelet eviction

Option 2 (my preference):
- Scheduling, Preemption and Eviction:
  - Other scheduling topics...
  - Pod Disruption
    - Preemption
    - User initiated Eviction
    - Kubelet eviction

Once we have figured the concept hierarchy, we can find better names (like node-pressure eviction as opposed to kubelet eviction).

sftim commented:
I'm fine with either layout from https://github.com/kubernetes/website/pull/27739#issuecomment-838931700

If we pick the one with a “Pod Disruption” section, that section index should _definitely_ mention PodDisruptionBudget (eg by hyperlinking).

ahg-g reviewed (commented):
> React with 👍 to _this_ message if you prefer:
> 
> > Option 2:
> > 
> > * Scheduling, Preemption and Eviction:
> >   
> >   * Other scheduling topics...
> >   * Pod Disruption
> >     
> >     * Preemption
> >     * User initiated Eviction
> >     * Kubelet eviction

if we want to call it kubelet eviction, then lets name the first one "Scheduler preemption"

shannonxtreme commented on content/en/docs/concepts/scheduling-eviction/pod-eviction.md:
```diff
@@ -0,0 +1,432 @@
+---
+
+Eviction is the process by which {{<glossary_tooltip text="pods" term_id="pod">}}
+in a {{<glossary_tooltip text="node" term_id="node">}} are
+terminated. 
```
We explain the nature of the termination in the details (and terminated sounds forceful).

shannonxtreme replied:
We use _involuntary disruption_ to describe kubelet evictions in https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#voluntary-and-involuntary-disruptions, so maybe we call this "involuntarily terminate"?

ahg-g replied:
hmm, there are other types of involuntary disruption not mentioned here (like the VM itself getting evicted), I am fine keeping it as is.

sftim replied:
If feasible, [rewrite this phrase sentence in the active voice ](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice)- this means we state what forces the termination.

shannonxtreme replied:
Doesn't apply anymore after the refactoring, but you may have suggestions for the new glossary entry.

shannonxtreme commented on content/en/docs/reference/glossary/kubelet-eviction.md:
```diff
@@ -0,0 +1,24 @@
+---
+title: Node-pressure eviction
+id: kubelet-eviction
```
Will do, this was me being lazy because kubelet eviction was easier to anchor to :P

shannonxtreme commented:
I think Descheduling should either be a part of the Scheduler concept topic or grouped under a Scheduling section alongside the other Scheduler topics in this folder, similar to how Pod Disruption will group the eviction and preemption stuff.

alculquicondor commented:
Re: descheduling

I say that's yet another form of disruption. @damemi?

ahg-g commented on content/en/docs/reference/glossary/api-eviction.md:
```diff
@@ -0,0 +1,22 @@
+---
```
ah, it is under the "operation" filter.

sftim commented:
(so long as we've thought about a home for Descheduler, I don't really mind where it lives)

Huang-Wei commented:
maybe add a general item like "custom eviction integrators" under Pod Disruption? so that we can mention something like: thanks for the extensibility of the Kubernetes system, the evictions behavior can be extended by custom solutions, such as de-scheduler, blabla...

shannonxtreme commented:
Alright. @sftim, @alculquicondor and @Huang-Wei, I refactored the docs like we discussed. 

- Added new glossary entry for Disruption
- Split eviction into API and Node-pressure
- Redirects
- Scheduling main page has grouping

Please let me know if the groups need to be changed. Also, I will give the docs a final look for links and stuff after we've got your LGTMs.

alculquicondor commented on content/en/docs/concepts/scheduling-eviction/_index.md:
```diff
@@ -1,11 +1,39 @@
+* [Node-pressure Eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/)
+* [API-initiated Eviction](/docs/concepts/scheduling-eviction/api-eviction/)
+
+<!-- TODO: Add Descheduling under Disruption when creating topic -->
```
just remove this TODO. There are no plans to do this at the moment.

alculquicondor commented on content/en/docs/reference/glossary/api-eviction.md:
```diff
@@ -0,0 +1,22 @@
+using a client of the kube-apiserver, like the `kubectl drain` command. 
+When an `Eviction` object is created, the API server terminates the Pod. 
+
+API-initiated eviction is not the same as [node-pressure eviction](/docs/concepts/scheduling-eviction/eviction/#kubelet-eviction).
```
old link?

shannonxtreme replied:
Yes, good catch. I'll go through the doc after lgtms to find other sneaky links too. Thanks!

alculquicondor commented on content/en/docs/concepts/scheduling-eviction/pod-priority-preemption.md:
```diff
@@ -4,7 +4,7 @@ reviewers:
+weight: 50
 ---
 
 <!-- overview -->
```
Add mentions of the other types of disruption? It's fine for a follow up PR.

shannonxtreme replied:
I linked to it in a whatsnext section.

alculquicondor commented on content/en/docs/concepts/scheduling-eviction/_index.md:
```diff
@@ -1,11 +1,39 @@
+* [Scheduler Performance Tuning](/docs/concepts/scheduling-eviction/scheduler-perf-tuning/)
+* [Resource Bin Packing for Extended Resources](/docs/concepts/scheduling-eviction/resource-bin-packing/)
+
+## Pod Disruption
```
Add a short description?

shannonxtreme replied:
The glossary definition acts as the short description in this case (see staging at https://deploy-preview-27739--kubernetes-io-master-staging.netlify.app/docs/concepts/scheduling-eviction/#pod-disruption)

alculquicondor replied:
nit: maybe we could add an extra line saying something like:

`These are the types of disruptions the control plane can apply to pods:`

And should we reorder `Preemption` to be before `Priority`? 🤔

shannonxtreme replied:
I think instead we call the link Pod Preemption and link to the Preemption section of that page?

alculquicondor replied:
sgtm

alculquicondor replied:
no wait, but this is the index, it should have the full name.

alculquicondor commented on content/en/docs/concepts/scheduling-eviction/api-eviction.md:
```diff
@@ -0,0 +1,16 @@
+
+{{< glossary_definition term_id="api-eviction" length="short" >}} </br>
+
+You can request eviction by directly calling the Eviction API 
```
Link to https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api ?

shannonxtreme replied:
It is linked in the glossary definition (so looks OK in staging at: https://deploy-preview-27739--kubernetes-io-master-staging.netlify.app/docs/concepts/scheduling-eviction/api-eviction/)

alculquicondor replied:
looks good.

Now, I think some of the contents of https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api can be put here as well. Note that that doc is a "task", meaning that it has a tutorial structure. In "concepts", we should make it more descriptive, explaining the fields and responses. But this is definitely for a follow up.

shannonxtreme replied:
I'll create a new issue for this after we publish this doc

alculquicondor commented on content/en/docs/concepts/scheduling-eviction/api-eviction.md:
```diff
@@ -0,0 +1,16 @@
+API-initiated evictions respect your configured [`PodDisruptionBudgets`](/docs/tasks/run-application/configure-pdb/)
+and [`terminationGracePeriodSeconds`](/docs/concepts/workloads/pods/pod-lifecycle#pod-termination). 
+
+API-initiated evictions are not the same as [node-pressure eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/). 
```
Add a "whatsnext" section linking to the other two types of eviction.

alculquicondor commented on content/en/docs/concepts/scheduling-eviction/api-eviction.md:
```diff
@@ -0,0 +1,21 @@
+API-initiated evictions respect your configured [`PodDisruptionBudgets`](/docs/tasks/run-application/configure-pdb/)
+and [`terminationGracePeriodSeconds`](/docs/concepts/workloads/pods/pod-lifecycle#pod-termination). 
+
+API-initiated evictions are not the same as [node-pressure eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/). 
```
this line seems unnecessary given the next section

alculquicondor commented on content/en/docs/reference/glossary/node-pressure-eviction.md:
```diff
@@ -0,0 +1,24 @@
+aka: kubelet eviction
+tags:
+- operation
+- fundamental
```
is it?

sftim replied:
I'd omit “fundamental”

alculquicondor reviewed (commented):
overall lgtm from scheduling perspective

alculquicondor reviewed (commented):
/lgtm
I'm fine with this as an iteration.

sftim commented:
https://github.com/kubernetes/website/compare/cc4d451f3a75fb53783a81d5bd6753f6fedbc01b..c184f20f76eaf58ece0ef14e025da1a4fa9d4f0d is fine by me
/lgtm

shannonxtreme commented:
👀

shannonxtreme commented:
@sftim what else do I need for approval on this one?

sftim commented:
If someone who's not me LGTMs this, I can approve it.

shannonxtreme commented:
@alculquicondor sorry, I pushed to implement your final comment and it removed your lgtm. Would you mind adding it again? Thanks!

alculquicondor commented:
/lgtm

still applies :)

sftim commented:
Let's do this
/approve
